### PR TITLE
Pid dev

### PR DESCRIPTION
**Pull request by jbellevi**
Pre-competition merge into master

### CODE REVIEW

File: nodes/control_utils.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import numpy as np
```
Comment (jbellevi) [inline nodes/control_utils.py]: Remove if no longer relevant

File: nodes/image_processing_utils.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
+import cv2
+import numpy as np
+import copy
+
+
+def get_hough_lines(img, img_type="edges", threshold=125, probabilistic=False, min_line_length=100, max_line_gap=10):
+    """
+
```
Comment (jbellevi) [inline nodes/image_processing_utils.py:9]: Define hough line / add explanation

File: nodes/image_processing_utils.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    @param img: image to generate lines on
+    @param threshold: threshold to pass to the Houghlines generator
+    @param probabilistic: if true uses the probabalistic hough transform see cv2.HoughLinesP (
+    @param min_line_length: minimum length of a line, passed to cv2.HoughLinesP
+    @param max_line_gap: maximum gap in a line, passed to cv2.HoughLinesP
+    @param verbose: Whether to print information to the console (Currently unimplemented)
```
Comment (jbellevi) [inline nodes/image_processing_utils.py]: remove if unimplemented

File: nodes/image_processing_utils.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        lines = cv2.HoughLinesP(edges, 1, np.pi / 180,
+                                threshold, min_line_length, max_line_gap)
+
+    else:
+        lines = cv2.HoughLines(edges, 1, np.pi / 180, threshold)
+
+    return lines
+
+
+def draw_lines(img, lines, probabilistic=False, color=(0, 0, 255), size=2):
```
Comment (jbellevi) [inline nodes/image_processing_utils.py]: Quick @param that you're drawing line to the image?

File: nodes/image_processing_utils.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                x1 = int(x0 + length * (-b))
+                y1 = int(y0 + length * a)
+                x2 = int(x0 - length * (-b))
+                y2 = int(y0 - length * a)
+
+                cv2.line(img, (x1, y1), (x2, y2), color, size)
+
+
+def find_cardinal_clusters(angles):
+    # The K-means algorithm is sometimes unreliable.
```
Comment (jbellevi) [inline nodes/image_processing_utils.py]: Clarify

File: nodes/image_processing_utils.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        for angle in angles:
+
+            if theta_0 - half_bucket <= angle[0] <= theta_0 + half_bucket:
+                buckets[0].append(angle)
+            elif theta_0 - half_bucket - np.pi / 2.0 <= angle[0] <= theta_0 + half_bucket - np.pi / 2.0:
+                buckets[1].append(angle)
+            elif theta_0 - half_bucket + np.pi / 2.0 <= angle[0] <= theta_0 + half_bucket + np.pi / 2.0:
+                buckets[2].append(angle)
+
+        include = buckets[0].__len__() + buckets[1].__len__()
```
Comment (jbellevi) [inline nodes/image_processing_utils.py:62]: len(buckets[0]) == buckets[0].__len__() is more regular form

File: nodes/letter_isolator_beta.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python
+
+import cv2
+import numpy as np
+
+
+class LetterIsolatorBeta:
```
Comment (StephenJamesCashen) [inline nodes/letter_isolator_beta.py:6]: Rename from Beta

File: nodes/letter_isolator_beta.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        while i in range(1, len(letter_rect)):
+            if (abs(letter_rect[i][0] - letter_rect[i - 1][0])
+                    <= width // 12):
+                if (letter_rect[i][2] * letter_rect[i][3] > letter_rect[i - 1][2] * letter_rect[i - 1][3]):
+                    letter_rect[i - 1] = letter_rect[i]
+                del letter_rect[i]
+            else:
+                i += 1
+
+        # if self.is_testing:
```
Comment (jbellevi) [inline nodes/letter_isolator_beta.py]: remove giant blocks of testing

File: nodes/plate_isolator_colour.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/env python
+
+import cv2
+import numpy as np
+from scipy.spatial import distance as dist
+
+
+def _contour_length_tuple(c):
```
Comment (jbellevi) [inline nodes/plate_isolator_colour.py]: Check: do I still use this?

File: nodes/letter_isolator_beta.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        @param img_width: width that input image gets scaled to
+        """
+        self.bin_thresh_plate = bin_thresh_plate
+        self.is_testing = testing
+
+    def get_chars(self, parking, license):
+        """
+        Given an image of a parking and license plate,
+        returns each character as a separate image.
+        @param img: greyscale image of the side of the car (with both plates)
```
Comment (StephenJamesCashen) [inline nodes/letter_isolator_beta.py]: Img isn't a parameter. Missing parameters for parking and license

File: nodes/image_processing_utils.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    tolerance = np.pi / 60  # 3 degrees
+
+    head_abs = np.abs(headings[1])
+    if np.abs(head_abs - 1.57) < tolerance:
+        headings[1] = turn_bias * head_abs
+
+    return headings, clusters
+
+
+def find_crosswalk(img):
```
Comment (jbellevi) [inline nodes/image_processing_utils.py:122]: Add some comment: less about functionality and more about what it DOES once it finds the crosswalk (or how it returns it)

File: nodes/plate_isolator_colour.py [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    return cv2.arcLength(c[0], True)
+
+
+class PlateIsolatorColour:
+    """
+    The goal of this module is to pick out parking
+    1. will pull out cars by colour
+
+    Input: a clean image of the plates
+           Random, likely terrible images picked up from Anki camera
```
Comment (jbellevi) [inline nodes/plate_isolator_colour.py]: Update this!!

File: nodes/letter_isolator_beta.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python
+
+import cv2
+import numpy as np
+
+
+class LetterIsolatorBeta:
+    """
+    Given coloured image of the parking and license plates
+    Returns the binarised isolated letters
```
Comment (jbellevi) [inline nodes/letter_isolator_beta.py:9]: Is it still? check

File: nodes/letter_isolator_beta.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+        letters = [None] * len(letter_rect)
+        for i, rect in enumerate(letter_rect):
+            # print(rect)
+            letters[i] = img[max(0, rect[1] - 2):min(img.shape[0] - 1, rect[1] + rect[3] + 2),
+                             max(0, rect[0] - 2):min(img.shape[1] - 1, rect[0] + rect[2] + 2)]
+        letters_new = [self._scale_and_blur(l) for l in letters]
+        return letters_new
+
+    def get_chars_from_plate_hardcode(self, img, expected_letters):
```
Comment (jbellevi) [inline nodes/letter_isolator_beta.py]: Add docstring

File: nodes/letter_isolator_beta.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+        img = cv2.threshold(grey, self.bin_thresh_plate, 255,
+                            cv2.THRESH_BINARY)[1]
+        return cv2.morphologyEx(img, cv2.MORPH_OPEN, (6, 6))
+
+    def display_test(self, img, duration=3000):
+        if (self.is_testing):
+            cv2.imshow("testing", img)
+            cv2.waitKey(duration)
+
+    def get_chars_from_plate(self, img, expected_letters=2, thresh=200):
```
Comment (jbellevi) [inline nodes/letter_isolator_beta.py]: remove if not in use

File: nodes/letter_isolator_beta.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+        Binarises the image to isolate plates
+        @param: input image
+        @return: the binarised image (plate black, rest white)
+        """
+        grey = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+        img = cv2.threshold(grey, self.bin_thresh_plate, 255,
+                            cv2.THRESH_BINARY)[1]
+        return cv2.morphologyEx(img, cv2.MORPH_OPEN, (6, 6))
+
+    def display_test(self, img, duration=3000):
```
Comment (jbellevi) [inline nodes/letter_isolator_beta.py]: docstring

File: nodes/letter_isolator_beta.py [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+            cv2.imshow("testing", img)
+            cv2.waitKey(duration)
+
+    def get_chars_from_plate(self, img, expected_letters=2, thresh=200):
+        """
+        Returns an ordered list of letters imgs on plate (left to right)
+        @param img - image from which letters are extracted
+        @param expected_letters - number of expected letters
+                                  (2 for parking, 4 for license plate)
+
```
Comment (StephenJamesCashen) [inline nodes/letter_isolator_beta.py:78]: missing thresh parameters

File: nodes/letter_isolator_beta.py [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+            n0 = self._scale_and_blur(img[y_top:y_bottom,
+                                      w // 2: w - x_offset_inner_right])
+            n1 = self._scale_and_blur(img[y_top:y_bottom,
+                                      w - x_offset_inner_right:
+                                      w - x_offset_outer])
+            return c0, c1, n0, n1
+        else:
+            raise ValueError("Expected number of letters must be 2 or 4")
+
+    def _scale_and_blur(self, img):
```
Comment (jbellevi) [inline nodes/letter_isolator_beta.py:113]: Add docstring

File: nodes/letter_isolator_beta.py (hunk 8 above)
Comment (StephenJamesCashen) [inline nodes/letter_isolator_beta.py]: Block comments

File: nodes/parking_bot.py [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+    Compares plates found to plates generated.
+    :param discovered_plates: a list of plates discovered by the parking bot.
+    :return: True if all plates discovered are in the generated plates csv.
+    """
+    generated_plates = pandas.read_csv(gv.path + "/../scripts/plates.csv")
+    return all(plate in generated_plates for plate in discovered_plates)
+
+
+def get_image(imgmsg):
+    # Try to convert
```
Comment (jbellevi) [inline nodes/parking_bot.py]: complete sentence

File: nodes/letter_isolator_beta.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+        #     cv2.imshow("bound rect", copy)
+        #     print(boundRect)
+        #     cv2.waitKey(2000)
+
+        assert len(letter_rect) == expected_letters, \
+            "letter_rect length: {}, {}".format(len(letter_rect), letter_rect)
+
+        letters = [None] * len(letter_rect)
+        for i, rect in enumerate(letter_rect):
+            # print(rect)
```
Comment (StephenJamesCashen) [inline nodes/letter_isolator_beta.py]: useless comment

File: nodes/parking_bot.py [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+    cv_image = bridge.imgmsg_to_cv2(imgmsg)
+    cv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2RGB)
+    return cv_image
+
+
+class ParkingBot:
+
```
Comment (jbellevi) [inline nodes/parking_bot.py:31]: gimme an overview docstring

File: nodes/letter_isolator_beta.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+            letters[i] = img[max(0, rect[1] - 2):min(img.shape[0] - 1, rect[1] + rect[3] + 2),
+                             max(0, rect[0] - 2):min(img.shape[1] - 1, rect[0] + rect[2] + 2)]
+        letters_new = [self._scale_and_blur(l) for l in letters]
+        return letters_new
+
+    def get_chars_from_plate_hardcode(self, img, expected_letters):
+        w = img.shape[1]
```
Comment (StephenJamesCashen) [inline nodes/letter_isolator_beta.py:84]: docstring if you choose

File: nodes/parking_bot.py [hunk 23]
```diff
@@ -0,0 +1,405 @@
...
+        # When was the last time we checked for movement
+        self.last_difference_check_time = None
+        # Is the robot currently stopping for a crosswalk
+        self.stopping_for_crosswalk = False
+        # Time we left the crosswalk
+        self.left_cross_walk_time = None
+        # Is the bot waiting for a pedestrian to begin moving
+        self.wait_for_movement_on_crosswalk = None
+
+    def process_image(self, imgmsg):
```
Comment (jbellevi) [inline nodes/parking_bot.py]: docstring

File: nodes/plate_isolator_colour.py [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+                ([30, 0, 0], [40, 255, 255])
+            ]
+        else:
+            self.colour_bounds = colour_bounds
+
+        self.testing = testing
+
+    def extract_plates(self, img, duration=1000):
+        """
+        Returns plates in order: parking, license, or None if no plates found
```
Comment (StephenJamesCashen) [inline nodes/plate_isolator_colour.py]: docstring

File: nodes/plate_isolator_colour.py [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+            cv2.imshow("image", img)
+            cv2.imshow("parking", parking)
+            cv2.imshow("license", license)
+            cv2.waitKey(duration)
+            # print("success")
+
+        return parking, license
+
+    def get_car_mask(self, img, duration=3000):
+        bound_num = 0
```
Comment (StephenJamesCashen) [inline nodes/plate_isolator_colour.py:106]: docstring

File: nodes/plate_isolator_colour.py [hunk 26]
```diff
@@ -0,0 +1,310 @@
...
+
+    def get_plate_corners(self, img, mask, colour):
+        """
+        Returns 4 corners of the plates (with perspective still) as an np array
+        Parking plate first, then license plate
+        @param img - img in which we search for plate
+        @param mask - the mask of the car
+        @param colour - the colour of the car (that will also be part of
+                        contour and thus should be filtered)
+
```
Comment (StephenJamesCashen) [inline nodes/plate_isolator_colour.py:157]: return?

File: nodes/parking_bot.py [hunk 27]
```diff
@@ -0,0 +1,405 @@
...
+                                                      self.parking_dict[parking][1]))
+                    print('New {}: {} ({})'.format(parking, license_text, prob[2]))
+                if self.parking_dict[parking][1] < prob[2]:
+                    self.parking_dict[parking] = (license_text, prob[2])
+
+        probabilities = list(self.parking_dict[parking][1] for parking in self.parking_dict)
+        """
+        If we have read 6 license plates and are confident in their readings we finish the run.
+        """
+        if self.parking_dict.__len__() == 6 and min(probabilities) > 3.99:
```
Comment (jbellevi) [inline nodes/parking_bot.py]: magic number (justify?)

File: scripts/Homography Generation/detector.py [hunk 28]
```diff
@@ -0,0 +1,81 @@
+"""
+Playing with opencvs blob detector parameters for finding our top down homography.
```
Comment (jbellevi) [inline scripts/Homography Generation/detector.py]: I need answers

File: nodes/plate_reader.py [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+                                     tens_digit, prob_tens_digit,
+                                     ones_digit, prob_ones_digit)
+                else:
+                    self.parking_license_pairs[parking_num].\
+                        update(letter_left, prob_letter_left,
+                               letter_right, prob_letter_right,
+                               tens_digit, prob_tens_digit,
+                               ones_digit, prob_ones_digit)
+
+                # print("Parking {} ({})({})".format(parking_num,
```
Comment (StephenJamesCashen) [inline nodes/plate_reader.py]: Block comment

File: nodes/plate_reader.py [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+                        prob_ones_digit < self.certainty_thresh):
+                    return None, None, None
+
+                return parking_num, license_text, certainty
+            except AssertionError as e:
+                print(e)
+                print("letters could not be extracted from image")
+                return None, None, None
+        else:
+            # print("No plates found in image")
```
Comment (StephenJamesCashen) [inline nodes/plate_reader.py]: uselesscomment

File: scripts/Homography Generation/detector.py [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+im_with_keypoints = cv2.drawKeypoints(img, keypoints, np.array([]),
+                                      (0, 0, 255), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+
+cv2.imwrite("blobs1.jpg", im_with_keypoints)
+
+img = cv2.imread("sim-calibration.png")
+print(img.shape)
+gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+ret, thresh = cv2.threshold(gray, 200, 255, cv2.THRESH_BINARY)
+# cv2.imshow("thresh", thresh)
```
Comment (jbellevi) [inline scripts/Homography Generation/detector.py]: remove commented out code

File: nodes/plate_reader.py [hunk 32]
```diff
@@ -0,0 +1,231 @@
...
+    parking_spot, text, _ = plate_reader.process_image(img)
+    # if (parking_spot is not None):
+    #     print("Parking: {}, {}".format(parking_spot, text))
+
+
+def main():
+    path = 'Preprocessing/IsolatingPlates/images_new_format/'
+    plate_reader = PlateReader('num_model.h5', 'char_model_new.h5')
+    for i in range(69, 73):
+        # print("image {}".format(i))
```
Comment (StephenJamesCashen) [inline nodes/plate_reader.py]: useless comment

File: scripts/Homography Generation/generate_homography.py [hunk 33]
```diff
@@ -0,0 +1,101 @@
...
+img_1 = cv2.imread("sim-calibration.png")
+
+gray_img = cv2.cvtColor(img_1, cv2.COLOR_BGR2GRAY)
+ret, thresh = cv2.threshold(gray_img, 200, 255, cv2.THRESH_BINARY)
+
+img = cv2.imread("sim-calibration.png")
+print(img.shape)
+gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+ret, thresh = cv2.threshold(gray, 200, 255, cv2.THRESH_BINARY)
+# cv2.imshow("thresh", thresh)
```
Comment (jbellevi) [inline scripts/Homography Generation/generate_homography.py]: kill it

File: nodes/plate_reader.py [hunk 34]
```diff
@@ -0,0 +1,231 @@
...
+    def __str__(self):
+        total_cert = self.c0[1] + self.c1[1] + self.n0[1] + self.n1[1]
+        return '{}{}{}{} ({})'.format(self.c0[0], self.c1[0], self.n0[0],
+                                      self.n1[0], total_cert)
+
+
+def test(path, plate_reader):
+    img = cv2.imread(path)
+    parking_spot, text, _ = plate_reader.process_image(img)
+    # if (parking_spot is not None):
```
Comment (StephenJamesCashen) [inline nodes/plate_reader.py]: Useless comment

**Pull request merged**